ETDRK integrators

MSeeker1340 opened the pull request:
This PR will contain a series of updates to update and implement the "classical" ExpRK integrators. These methods do not use internal time-stepping for Krylov, and are casted in the semilinear formulation. 

First some minor updates on the interface part. Note that for `integrator.opts.internalnorm`, the default choice (`ODE_DEFAULT_NORM`) does not support `norm(A, Inf)`, so I have to switch to using `Base.norm` in the tests. Might be a good idea to just change `norm(A, Inf)` in `arnoldi!` to `norm(A)` though, considering that we will be using fast normest for large sparse `A` anyways.

ChrisRackauckas commented:
>Might be a good idea to just change norm(A, Inf) in arnoldi! to norm(A) though, considering that we will be using fast normest for large sparse A anyways.

`norm(A)` would be the L2 norm which needs an SVD factorization though?

ChrisRackauckas commented:
>The in-place version is a bit trickier because we need to allocate cache for the computation of the Jacobian, but currently there's no way to infer the type of the Jacobian used by f.jac(J,u,p,t). 

For now assume it's dense. We'll have to just overhaul this all at the same time.

MSeeker1340 commented:
Got stuck with coding for a moment so I decided to go back to sort out the algebra. Some notes to help me organize things:

# Classical ExpRK methods

For an autonomous system $u' = F(u) = Au + g(u)$, an $s$-stage method has the form:

$$ u_{n+1} = e^{hA}u_n + h \sum_{i=1}^sb_i(hA)G_{ni}$$

$$ U_{ni} = e^{c_ihA}u_n + h \sum_{j=1}^sa_{ij}(hA)G_{nj}$$

$$ G_{nj} = g(t_n + c_jh, U_{nj}) $$

However since the $b_i$ and $a_{ij}$ need to satisfy the condition (2.21) in the Hochbruck review, the $e^{c_ihA}$ factors can be absorbed. Also for explicit methods we have $a_{ij} = 0$ if $ i \le j$. Therefore we can cast the method in a more convenient form:

$$ u_{n+1} = u_n + h\sum_{i=1}^sb_i(hA)(G_{ni} + Au_n) $$

$$ U_{ni} = u_n + h\sum_{j=1}^{i-1}a_{ij}(hA)(G_{nj} + Au_n) $$

# How to do Krylov

The naive approach of calculating $U_{ni}$ sequentially is not very efficient. To fully utilize the scaling property of the Kyrlov approximation, we should instead compute all products involving $G_{ni} + Au_n$ in one go (this is not unlike the "vertical" EPIRK methods, however here we can use general $b_i$ and $a_{ij}$ whereas in EPIRK the time-stepping Krylov requires functions in one column to all be some $\varphi_k$). That is, we first compute $F_{n1}$, then do `arnoldi!` with `A` on $F_{n1}$ and compute $b_1(hA)(G_{n1} + Au_n)$ and $a_{21}(hA)(G_{n1} + Au_n)$ through $a_{s1}(hA)(G_{n1} + Au_n)$; After this, compute $G_{n2}$ and follows suit.

Therefore, for an $s$-stage method only $s$ `arnoldi!` calls are needed.

# Implementation details

The most important part is the computation of $G_{nj}$. The others are just a combination of `arnoldi!`, `phiv!` and vector additions.

For an `integrator.f` given as a `SplitFunction`, both `A` and `g` are known and $G_{nj}$ can be calculated directly.

For an `integrator.f` given as a `DiffEqFunction` with `f.jac` known, the first step of `perform_step!` should be the calculation of `A`. For out-of-place style:

```julia
A = isa(f, SplitFunction) ? f.f1 : f.jac(uprev, p, t)
```

and for in-place style:

```julia
if isa(f, SplitFunction)
  A = f.f1
else
  f.jac(Jcache, uprev, p, t)
  A = Jcache
end
```

The allocation of `Jcache` should be taken care of by `alg_cache`, which might not be trivial as the type of the jacobian can vary.

For $G_{nj}$, one way is to mimic what is done for the Jacobian:

```julia
g = isa(f, SplitFunction) ? f.f2 : (u,p,t) -> f(u,p,t) - A*u
```

And similarly for the in-place style. What I'm worried about is if creating a new anonymous function in each `perform_step!` means the function needs to be compiled each time, which causes performance hit. I can also define a helper function `_compute_G` and has it dispatch on the type of `f`. Although less elegant, this should not cause performance issues but still serve to abstract away the computation of $G_{nj}$.

One thing I'm still not certain about is $G_{n1}$. It deserves special treatment because it is just $g(t_n, u_n)$, and has the potential to be FSALed. Indeed for `SplitFunction` its value is needed anyway to compute the derivative at the end of `perform_step!` (for interpolation), so it makes sense to FSAL its value (this is what is done for `LawsonEuler` and `NorsettEuler` currently, btw). However, for the generic case the $g$ function is different in each step and we don't have this privilege. Should we make a special case for the semilinear case or is this too complicated in view of potential performance gain?

MSeeker1340 commented:
So I decided to drop `ExpRKFsal` and switch back to FSALing just the right hand side. The rationales are:

1. As stated in the previous comment, the old approach does not work for generic `DiffEqFunction` and doens't really save much for `SplitFunction` (only one `g` function call per step, which is negligible especially for higher order methods).

2. This makes things simpler. #276 won't be an issue anymore and the structure of `LawsonEuler` & `NorsettEuler` will look a lot like plain old `Euler`, which can be helpful to future contributors who want to get into the ExpRK stuff.

Switching the FSAL type turned out to require major refactoring of the codebase. I have taken this opportunity to also implement some quality-of-life changes, such as consolidating the `initialize!` method for both `LawsonEuler` and `NorsettEuler`. Since I haven't touched `ETDRK4` yet I will leave it as-is for the moment. The end goal is for all ExpRK integrators, current and future ones, to have the same `initialize!` method. The framework for the coherent handling of `DiffEqFunction` and `SplitFunction` in the previous comment is also implemented (the `_compute_nl(!)` function).

The refactoring is far from complete though. Things I want to try:

- [ ] Consolidate the `alg_cache` method. Though different integrators do use different caches, most of the contents are the same and we should not need to repeat them.

- [ ] Modify the caching version of `NorsettEuler` to use the more convenient form (see previous comment). This will make it consistent with the Krylov version.

- [ ] Take a more critical look at the caches. Compared to `Euler`'s implementation, I have already removed `fsal` from the cache, which is completely unnecessary and redundant. The question I have now is whether `u` and `uprev` has a similar story. Think I need to consult `solve.jl` for this.

MSeeker1340 commented:
@ChrisRackauckas Not sure how you feel about using macro expansion in source files but I feel like it's the way to go if we want to minimize boilerplate code. 

The aim is to consolidate `alg_cache` and `perform_step` to the point that a new algorithm can be easily added simply defining its version of `expRK_operators`. This is partially done for constant caches. For the in-place style I'm not sure how far I can go but at least for precomputing the operators `expRK_operators` can also be of use.

An implementation detail is that I cannot set the type signature of `ops` to be `NTuple{N, opType}` and for the Krylov case use the empty tuple `()`. For some reason this is not accepted, as shown by the demo:

```julia
struct S{N, T}
    val::NTuple{N, T}
end
```


```julia
S((1,2,3))
```




    S{3,Int64}((1, 2, 3))




```julia
S(())
```


    MethodError: Cannot `convert` an object of type Tuple{} to an object of type S
    This may have arisen from a call to the constructor S(...),
    since type constructors fall back to convert methods.

    

    Stacktrace:

     [1] S(::Tuple{}) at .\sysimg.jl:77

ChrisRackauckas commented:
which source is ExpTrapezoid from?

MSeeker1340 commented:
> which source is ExpTrapezoid from?

It's described in the Hochbruck review paper (http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.296.2179&rep=rep1&type=pdf) in Example 2.6, which is a special case of the 2nd order method family in Example 2.18.

Incidentally this is also the ETD2RK scheme in Cox & Matthews' original ETD paper (https://pdfs.semanticscholar.org/9a47/a0c3f84a5aa1ef8016b8d9655951bc830933.pdf), although the original formulation looks very different.

MSeeker1340 commented:
By the way, one thing I haven't decided yet is what third and fourth order classical methods to implement. The more popular ETDRK3/ETDRK4 are only third and fourth order in the regular sense, but their worst case "stiff order" can be as low as 2.

Hochbruck & co spent a lot of effort in deriving stiffly accurate ExpRK integrators, which are described in the review paper. Unfortunately the schemes are not given names. Also the third and fourth order methods are all given as a one-parameter family. It's not clear at first sight what should be the best choice, unlike the case of ExpTrapezoid.

MSeeker1340 commented:
@ChrisRackauckas brought up a good point that the classical ETDRK methods might be more efficient in their original formula for small systems where the operators are precomputed and cached. So I decided to take a closer look at this and compare the formulas carefully.

Since we're comparing caching methods I shall assume the system is given with a `SplitFunction`, and that the computation cost of $g(t, u)$ is negligible compared to `A_mul_B!`.

# ETDRK1 (ETD1, Norsett-Euler)

ETD formulation (#matvecs = 2):

$$ u_{n+1} = \exp(hA)u_n + h\varphi_1(hA) G_n $$

Hochbruck formulation (#matvecs = 1):

$$ u_{n+1} = u_n + h\varphi_1(hA) (G_n + Au_n) $$

Note that we don't actually need to compute $Au_n$ as $G_n + Au_n$ can be carried over from the previous step via FSAL. The 

# ETDRK2 (ExpTrapezoid)

ETD formulation (#matvecs = 3):

$$ a_n = \exp(hA)u_n + h\varphi_1(hA)G_n $$
$$ u_{n+1} = a_n + h\varphi_2(hA)(g(t_n + h, a_n) - G_n) $$

Hochbruck formulation (#matvecs = 4):

$$ a_n = u_n + h\varphi_1(hA) (G_n + Au_n) $$
$$ u_{n+1} = u_n + h(b_1(hA)(G_n + Au_n) + b2(hA)(g(t_n + h,a_n) + Au_n)) $$

Where $b_1 = \varphi_1 - \varphi_2$ and $b_2 = \varphi_2$. The optimal method should be a combination of two: compute $a_n$ using Hochbruck (one matvec) and update $u$ using ETD (one matvec). Total #matvecs = 2.

# ETDRK3

ETD formulation (#matvecs = 7):

$$ a_n = \exp(hA/2)u_n + h/2\varphi_1(hA/2)G_n $$
$$ b_n = \exp(hA)u_n + h\varphi_1(hA)(2g(t_n + h/2, a_n) - G_n) $$
$$ u_{n+1} = \exp(hA)u_n + h(b_1(hA)f_1 + b_2(hA)f_2 + b_3(hA)f_3) $$

The last formula is rather long so I just grouped terms together as $b_1$, $f_1$, etc. The important part here is that it requires three matvecs, unlike the ETDRK2 case which can save one matvec by cancel out terms.

Hochbruck formulation (#matvecs = 6). To save space I shall just paste the Butcher table here:

![etdrk3](https://user-images.githubusercontent.com/30965945/41501436-59c782d4-7172-11e8-8ef7-d56df226d96a.PNG)

The matvecs needed are:

- One for computing $Au_n$;
- One for computing $a_n$ in the first stage;
- One for computing $b_n$ in the second stage (the two $\varphi_{1,3}$ can be grouped together);
- Three for the update of $u_{n+1}$.

# ETDRK4

ETD formulation (#matvecs = 9). Copied from Trefethen's paper:

![etdrk4](https://user-images.githubusercontent.com/30965945/41501502-abcebc04-7173-11e8-84ce-c9639d967473.PNG)

Hochbruck formulation (#matvecs = 8). Butcher table:

![etdrk4_butcher](https://user-images.githubusercontent.com/30965945/41501518-f5618dce-7173-11e8-9dfa-ef8909bf7152.PNG)

Breakdown:

- One for computing $Au_n$;
- One for computing $a_n$ in the first stage;
- One for computing $b_n$ in the second stage;
- Two for computing $c_n$ in the third stage;
- Three for the update of $u_{n+1}$ (note that $b_2$ and $b_3$ are the same).

# Conclusion

Except for ETDRK2, where a small trick in the original formulation can be used to save one matvec, the more modern Hochbruck formulation should be used.

MSeeker1340 commented:
One slight concern is that the testing script is taking longer and longer with every new method. I should probably remove a few tests and instead put the full test script in OrdinaryDiffEqExtendedTests.jl

ChrisRackauckas commented:
It's not pushing the time limit yet so we're still okay. Don't worry about it: I'm trying to get funding for a private CI to speed things up and will re-organize the tests around that.

MSeeker1340 commented:
ETDRK4 turns out to be more complicated than I thought. The main problem is the operator $a_{41}(hA) = \frac12\varphi_{1,3}(\varphi_{0,3} - 1)$ in the Butcher table, which is not a regular phi function. By treating it as the composition of two operators, we can compute this using an extra Krylov step.

However my naive implementation is failing tests at the moment, probably due to numerical instability from computing $\varphi_{0,3} - 1$ explicitly. I can try to rearrange the algebra and see if the problem can be avoided.

MSeeker1340 commented:
The problem for `ETDRK4` turns out to be typo all along... Nevertheless the numerical instability issue is still real and I fixed it for both the caching and Krylov version, the former via algebraic manipulation and the latter by going back to the ETD formula for the computation of the irregular term.

I also changed the testing script so that the test system no longer blows up, which I discovered while debugging. (It's actually quite remarkable that the previously implemented ExpRK integrators can actually handle solutions that are analytically unstable).

I think this PR is getting a little bit too large. Now that the ETD methods are complete it makes sense to merge this can do a separate one for the other classical ExpRK methods that I plan to include.

ChrisRackauckas commented:
I see more optimizing that can be done (and we should probably start benchmarking them), but I agree this PR is already big so we should merge (when tests pass) and continue in another PR.

MSeeker1340 commented on src/algorithms.jl:
```diff
@@ -649,20 +649,17 @@ struct GenericIIF2{F} <: OrdinaryDiffEqExponentialAlgorithm
-Base.@pure NorsettEuler(;krylov=false, m=30) = NorsettEuler(krylov, m)
 ETD1 = NorsettEuler # alias
 struct SplitEuler <: OrdinaryDiffEqExponentialAlgorithm end
 struct ETD2 <: OrdinaryDiffEqExponentialAlgorithm end
```
It's a multistep method. While the implementation might be simple I think it's better to consider exponential multistep methods as a whole. For now I shall focus on ExpRK.

ChrisRackauckas replied:
Got it. I agree. Let's keep multistep separate.

ChrisRackauckas commented:
Triggered a restart of Travis

MSeeker1340 commented:
@ChrisRackauckas OK to merge this now?
